### PR TITLE
Establish etcd/client APIs

### PR DESCRIPTION
**Pull request by bcwaldon**

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+		return
+	}
+
+	if leps != n {
+		err = errors.New("unable to pick endpoint: copy failed")
+		return
+	}
+
+	for _, ep := range eps {
+		hc := c.clientFactory(ep)
```
Comment (xiang90) [inline client/client.go]: why do we want to create a separate client for each request we are going to send?

File: client/client.go [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+func (c *httpClusterClient) Do(ctx context.Context, act httpAction) (resp *http.Response, body []byte, err error) {
+	c.RLock()
+	leps := len(c.endpoints)
+	eps := make([]url.URL, leps)
+	n := copy(eps, c.endpoints)
+	c.RUnlock()
+
+	if leps == 0 {
+		err = ErrNoEndpoints
+		return
```
Comment (xiang90) [inline client/client.go]: really do not like naked return

File: client/client.go [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+	case <-ctx.Done():
+		c.transport.CancelRequest(req)
+		// wait for request to actually exit before continuing
+		<-rtchan
+		err = ctx.Err()
+	}
+
+	// always check for resp nil-ness to deal with possible
+	// race conditions between channels above
+	defer func() {
```
Comment (xiang90) [inline client/client.go]: what happens if line 269 gets hit and then the defer func and then line 258? seems like the resp will remain opening.

File: client/client.go [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+	return resp, body, err
+}
+
+type redirectFollowingHTTPClient struct {
+	client        httpClient
+	checkRedirect CheckRedirectFunc
+}
+
+func (r *redirectFollowingHTTPClient) Do(ctx context.Context, act httpAction) (*http.Response, []byte, error) {
+	for i := 0; ; i++ {
```
Comment (xiang90) [inline client/client.go]: i hope we have a limited number for the loop

Comment (bcwaldon) [inline client/client.go]: It's up to the CheckRedirectFunc. DefaultCheckRedirect uses 10. Are you suggesting we pick an arbitrarily large number here?

Comment (xiang90) [inline client/client.go]: i want to detect the loop. so give it a large enough number is fine. maybe 100?

File: client/client.go [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+}
+
+type redirectedHTTPAction struct {
+	action   httpAction
+	location url.URL
+}
+
+func (r *redirectedHTTPAction) HTTPRequest(ep url.URL) *http.Request {
+	orig := r.action.HTTPRequest(ep)
+	orig.URL = &r.location
```
Comment (yichengq) [inline client/client.go]: I think it is a little weird. If it redirects 100 times, it will call 100 times to go deep into the original action.

File: client/doc.go [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+	import (
+		"net/http"
+
+		"github.com/coreos/etcd/client"
+		"golang.org/x/net/context"
+	)
+
+	cfg := client.Config{
+		Endpoints: []string{"http://127.0.0.1:4001"},
+		Transport: DefaultTransport,
```
Comment (yichengq) [inline client/doc.go]: http.DefaultTransport

File: client/keys.go [hunk 7]
```diff
@@ -28,57 +27,227 @@ import (
...
+	ErrorCodeIndexNaN          = 203
+	ErrorCodeInvalidField      = 209
+	ErrorCodeInvalidForm       = 210
+
+	ErrorCodeRaftInternal = 300
+	ErrorCodeLeaderElect  = 301
+
+	ErrorCodeWatcherCleared    = 400
+	ErrorCodeEventIndexCleared = 401
+)
```
Comment (yichengq) [inline client/keys.go]: why do we paste all error codes here instead of using error/error.go?

File: client/keys.go [hunk 8]
```diff
@@ -28,57 +27,227 @@ import (
...
+	// Next blocks until an etcd event occurs, then returns a Response
+	// represeting that event. The behavior of Next depends on the
+	// WatcherOptions used to construct the Watcher. Next is designed to
+	// be called repeatedly, each time blocking until a subsequent event
+	// is available.
+	//
+	// If the provided context is cancelled, Next will return a non-nil
+	// error. Any other failures encountered while waiting for the next
+	// event (connection issues, deserialization failures, etc) will
+	// also result in a non-nil error.
```
Comment (yichengq) [inline client/keys.go]: if it fails, what does the client expect to do? create a new watcher or call next again?

File: client/keys.go [hunk 9]
```diff
@@ -124,32 +309,46 @@ func (k *httpKeysAPI) Get(ctx context.Context, key string) (*Response, error) {
...
+	act := waitAction{
+		Prefix: k.prefix,
+		Key:    key,
+	}
+
+	if opts != nil {
+		act.Recursive = opts.Recursive
+		if opts.AfterIndex > 0 {
+			act.WaitIndex = opts.AfterIndex + 1
+		}
```
Comment (yichengq) [inline client/keys.go]: client cannot wait on 1 then, right? What is it supposed to do if it wants to wait from the head?

File: client/members.go [hunk 10]
```diff
@@ -23,31 +23,92 @@ import (
...
+	}{
+		PeerURLs: make([]string, len(m.PeerURLs)),
+	}
+
+	for i, u := range m.PeerURLs {
+		s.PeerURLs[i] = u.String()
+	}
+
+	return json.Marshal(&s)
+}
```
Comment (yichengq) [inline client/members.go]: why do we copy httptypes stuff here? I thought it is extracted to be shared by etcdhttp and client.

**Comment by yichengq** [comment]
I like the architecture overall.

File: client/client.go (hunk 2 above)
Comment (bcwaldon) [inline client/client.go]: This function grew quite a bit - much further than I anticipated. I'll get rid of the naked return.

File: client/client.go (hunk 1 above)
Comment (bcwaldon) [inline client/client.go]: To keep the code simple while it is still developing. I'm more than happy to cache a set of clients locally once we make sure they are threadsafe, etc.

File: client/client.go (hunk 4 above)
Comment (bcwaldon) [inline client/client.go]: Ok.

File: client/client.go (hunk 5 above)
Comment (bcwaldon) [inline client/client.go]: I don't quite understand what you're asking of me here.

File: client/client.go (hunk 1 above)
Comment (xiang90) [inline client/client.go]: ok

File: client/client.go (hunk 2 above)
Comment (xiang90) [inline client/client.go]: thanks

File: client/doc.go (hunk 6 above)
Comment (bcwaldon) [inline client/doc.go]: http.DefaultTransport doesn't fulfill the CancelableTransport interface since it's typed as a RoundTripper. I've copied it verbatim to this package.

Comment (xiang90) [inline client/doc.go]: @bcwaldon There are proposals to add that Cancelable thing to another standard interface. I guess everybody is doing this ugly thing right now.

Comment (bcwaldon) [inline client/doc.go]: Yeah :(

File: client/keys.go (hunk 7 above)
Comment (bcwaldon) [inline client/keys.go]: I don't want to force users of the etcd/client pkg to import any more packages. Having to vendor etcd/client and etcd/error and etcd/etcdserver/httptypes is annoying.

File: client/keys.go (hunk 8 above)
Comment (bcwaldon) [inline client/keys.go]: To answer your question: yes. However, Watcher.Next might fail with a permanent error that renders it useless (ErrorCodeEventIndexCleared, ErrorCodeWatcherCleared). We can probably be a little more helpful to clients, but this is the bare minimum of what we should do.

File: client/keys.go (hunk 9 above)
Comment (bcwaldon) [inline client/keys.go]: That's true. I believe the "after" semantic is much more useful to support than using "from" and preserving the ability to watch for index 1. I'm open to discussion.

File: client/members.go (hunk 10 above)
Comment (bcwaldon) [inline client/members.go]: Same reason that I decided to copy the error codes. I don't want to make our users import any more than etcd/client.

File: client/client.go (hunk 3 above)
Comment (bcwaldon) [inline client/client.go]: Good catch, it does look like a race if the context times out the call. I'll get a fix and a test together.

File: client/client.go (hunk 5 above)
Comment (yichengq) [inline client/client.go]: whenever it redirects, it builds a new action
```
+			act = &redirectedHTTPAction{
+				action:   act,
+				location: *loc,
+			}
```
So it will call act.action.action.action.action......HTTPRequest(ep) if it does many redirections, right?

File: client/keys.go (hunk 7 above)
Comment (yichengq) [inline client/keys.go]: That makes it easy to diverge from the upstream package. But if we don't change the const, that should work.

Comment (bcwaldon) [inline client/keys.go]: There will be no changes to this list, only additions. And we aren't actively building the v2 API, so even the additions should be minimal over time. The benefit of one fewer import greatly outweigh the maintenance burden on our part.

File: client/client.go (hunk 5 above)
Comment (bcwaldon) [inline client/client.go]: It creates a new net/http.Request each time a redirect is attempted. It is not duplicating the action object.

File: client/client.go (hunk 3 above)
Comment (bcwaldon) [inline client/client.go]: https://github.com/bcwaldon/etcd/commit/d7000a2af512a9fa485fdd6e65b38674124615c7

File: client/client.go (hunk 2 above)
Comment (bcwaldon) [inline client/client.go]: https://github.com/bcwaldon/etcd/commit/74a2d26115374fd0dfbc613520617964bf3bdf90

File: client/client.go (hunk 4 above)
Comment (bcwaldon) [inline client/client.go]: https://github.com/bcwaldon/etcd/commit/9de0031d6113eb7cd587fa990eb5ff1c02f96421

**Comment by bcwaldon** [comment]
@yichengq @xiang90 I addressed everything I felt was actionable. Please check back over this.

File: client/client.go (hunk 5 above)
Comment (yichengq) [inline client/client.go]: When it is called in Do, it builds a redirectedHTTPAction wrapper on act when redirect happens, iirc. After three times, it becomes something like:
```
&redirectedHTTPAction{
  action: &redirectedHTTPAction{
    action: &redirectedHTTPAction{
      action: act,
      location: ,
    },
    location: ,
  },
  location: ,
}
```
I think there is no need to make it iterated.

Comment (bcwaldon) [inline client/client.go]: Oh, I see what you mean, now. I'll fix that. I was just being lazy with variable naming.

Comment (bcwaldon) [inline client/client.go]: @yichengq Fixed here: https://github.com/bcwaldon/etcd/commit/1d047d107f34ee42002f161224d5eb188b9563d3

**Comment by bcwaldon** [comment]
@xiang90 @yichengq this is ready

File: client/client.go (hunk 3 above)
Comment (xiang90) [inline client/client.go]: lgtm

File: client/client.go (hunk 4 above)
Comment (xiang90) [inline client/client.go]: lgtm

File: client/client.go (hunk 2 above)
Comment (xiang90) [inline client/client.go]: lgtm

File: client/client.go [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+
+	if leps != n {
+		return nil, nil, errors.New("unable to pick endpoint: copy failed")
+	}
+
+	var resp *http.Response
+	var body []byte
+	var err error
+
+	for _, ep := range eps {
```
Comment (xiang90) [inline client/client.go]: as we have discussed, i think a client should randomly choose an endpoint at the beginning and keep on using it until there is a failure or there is an cluster update. 

but i guess the goal of this pr is just to set up the interface?

**Comment by xiang90** [comment]
@bcwaldon I did not take a line by line review. I review the core part, and it lgtm. And we already discussed about the API. I still have a few concerns around endpoint picking and cluster updating. But it may out of the scope of this pr. it is already large enough.

File: client/client.go (hunk 11 above)
Comment (bcwaldon) [inline client/client.go]: Yes. I would like to do this in a separate PR.
